### PR TITLE
Water - Christabel

### PR DESCRIPTION
**Pull request by cescarez**
# Task List
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe in your own words what the Model is doing in Rails | The Model is a class that allows us to interact with the data/database.
Describe in your own words what the Controller is doing in Rails | The Controller is the central hub and sends requests for data to the Model, receives the data, and sends it to the view to be rendered to the user that sent the HTTP Request.
Describe in your own words what the View is doing in Rails | As stated above, the View gets rendered to HTML/CSS as the interface the user/client can view/
Describe an edge-case controller test you wrote | Wrote several tests for checking that a redirect happens to a "Task Not Found" page if an invalid task ID was received in the request/route params.
What is the purpose of using strong params? (i.e. the params method in the controller) | `params` allows us to use route params and access them in our controller method(s). Ex: get the task ID from the route params to search our database for the correct Task. Strong params are used as a security measure and ensures only the data we are expecting is read by our methods.
How are Rails migrations related to Rails models? | Migrations shape the schema, which itself describes the structure of the database. If the structure of the database needs to change (ex: rename a column), a new migration can be created to rename the column and change the schema.
Describe one area of Rails that are still unclear on | I initially wrote the toggle_complete requirement in Wave 4 as hardcoded buttons that sent a patch request with the requisite params. When is hardcoding requests acceptable (vs. defining a controller action and route)? Also, database method calls that get memoized .... purpose? .size vs .count?

### CODE REVIEW

File: app/controllers/tasks_controller.rb [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        render :show
+      end
+      return
+    end
+  end
+
+  def not_found
+    render :template => 'tasks/not_found', :status => :not_found
+    return
+  end
```
Comment (beccaelenzil) [inline app/controllers/tasks_controller.rb:101]: hrmmm, this didn't work for me. `tasks/not_found` gave me this error:

"This page isn’t working
localhost redirected you too many times."

File: app/controllers/tasks_controller.rb [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    task_id = params[:id]
+    @task = Task.find_by(id: task_id)
+
+    if @task.nil?
+      redirect_to task_not_found_path
+      return
+    else
+      render :new
+      return
+    end
```
Comment (beccaelenzil) [inline app/controllers/tasks_controller.rb:43]: Rather than including code here to say `render  :new` it is convention to have a separate edit view. Since this view largely overlaps with the new view, this is a great place to use partial views.

File: app/controllers/tasks_controller.rb [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    elsif @task.destroy
+      redirect_to tasks_path
+      return
+    else
+      redirect_to task_path(@task)
+      return
+    end
+  end
+
+  def toggle_complete
```
Comment (beccaelenzil) [inline app/controllers/tasks_controller.rb:77]: Consider how making two separate methods, for instance `mark_complete` and `mark_incomplete` would make this action idempotent.

**Review by beccaelenzil — COMMENTED** [review]
# Task List




## Major Learning Goals/Code Review





| Criteria                                           | yes/no, and optionally any details/lines of code to reference |
| -------------------------------------------------- | ------------------------------------------------------------- |
| At least 6 commits with meaningful commit messages | ✔️                                                            |
| Routes follow RESTful conventions                  | ✔️                                                            |
| Uses named routes (like `_path`)                   | ✔️                                                            |
| Creates Models and migrations                      | ✔️                                                            |
| Creates styled views                               | ✔️ Nice styles!                                                            |
| Handles errors like nonexistant tasks              | ✔️ This wasn't working correctly for me                                                           |
| Uses `form_with` to render forms in Rails          | ✔️                                                            |

## Functional Requirements/Manual Testing

| Functional Requirement                                                         | yes/no |
| ------------------------------------------------------------------------------ | ------ |
| Successfully handles index & show                                              | ✔️     |
| index & show tests pass                                                        | ✔️     |
| Successfully handles: New, Create                                              | ✔️     |
| New, Create tests pass                                                         | ✔️     |
| Successfully handles: Edit, Update                                             | ✔️ What you have works, but it is convention to create a separate `edit` view.     |
| Edit, Update tests pass with valid & invalid task ids                          | ✔️     |
| Successfully handles: Destroy, Task Complete                                   | ✔️     |
| Tests for Destroy & Task Complete include tests for valid and invalid task ids | ✔️     |

## Overall Feedback

Good job on Task List! It functions as expected. You hit the learning goals of creating styled views, using strong params, and thoroughly testing your code. You created your first rails app. Nice work! On your next project, I encourage you to make use of partial views, particularly for the edit/new views. Nice work making a custom route, action and view for `not_found`. The `not_found` path didn't work for me as expected -- perhaps we can take a look at our next 1:1. Keep up the hard work! 



| Overall Feedback                | Criteria                                                                                                                                                                                      | yes/no |
| ------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------ |
| Green (Meets/Exceeds Standards) | 5+ in Code Review && 6+ in Functional Requirements                                                                                                                                            |✔️
| Yellow (Approaches Standards)   | 3+ in Code Review && 5+ in Functional Requirements, or the instructor judges that this project needs special attention                                                                        |
| Red (Not at Standard)           | 0-2 in Code Review or 0-4 in Functional Reqs, or assignment is breaking/doesn’t run with less than 5 minutes of debugging, or the instructor judges that this project needs special attention |





## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality              | Yes? |
| -------------------- | ---- |
| Perfect Indentation  | ✅    |
| Descriptive/Readable | ✅    |
| Concise              | ✅    |
| Logical/Organized    | ✅    |
